### PR TITLE
Inference benchmarking

### PR DESCRIPTION
**Pull request by bcui19**
Below we include a script, .yamls, and a README for benchmarking deepspeed inference.

### CODE REVIEW

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                # Make sure we are not generating a fake batch with a EOS token
+                while True:
+                    batch = torch.randint(
+                        0,
+                        config.model.vocab_size - 1,
+                        size=(batch_size, input_length
+                             )).to(f'cuda:{torch.cuda.current_device()}')
+                    if tokenizer.convert_tokens_to_ids(eos_token) not in batch:
+                        break
+                batch = batch.to(torch.long)
```
Comment (dskhudia) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:58]: could you add torch.cuda.synchronize here?

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                            with get_precision_context(config.precision):
+                                inference_model.generate(
+                                    batch,
+                                    max_new_tokens=output_length,
+                                    use_cache=True)
+                        else:
+                            inference_model.generate(
+                                batch,
+                                max_new_tokens=output_length,
+                                use_cache=True)
```
Comment (dskhudia) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:72]: Also synchronize here too.

**Review by dskhudia — APPROVED** [review]

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                                                       config.tokenizer)
+    model.eval()
+
+    tokenizer = AutoTokenizer.from_pretrained(config.tokenizer_name)
+
+    inference_model = deepspeed.init_inference(model, config=inf_config)
+
+    # Checking if deepspeed casts dtypes correctly
+    for n, p in inference_model.named_parameters():
+        print('n is: ', n, p)
```
Comment (vchiley) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:39]: next line is ok, but here you're printing the entire tensor.

File: examples/llm/inference/benchmarking/README.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+The repository includes the benchmark_deepspeed.py script, along with associated `.yaml` files, to run benchmarking. The script takes a `.yaml` file as input and outputs the latency (in seconds) and tokens per second for each run. We average over `num_runs=5`, which is defined in the `.yaml` file. Additionally, we iterate over various `batch_sizes`, `input_lengths`, and `output_lengths` to produce varying throughput metrics.
+
+## Usage
+To use the `benchmark_deepspeed.py` script, you need to provide a `.yaml` file that specifies the model configuration and other parameters such as the path to the model checkpoint and the input data. You can modify the default `.yaml` files provided in the repository or create your own `.yaml` file.
+
+To run the benchmarking script, use the following command:
+
+`python benchmark_deepspeed.py config.yaml`
+
+To run the scripts on [The MosaicML platform](https://www.mosaicml.com/blog/mosaicml-cloud-demo) we've also included scripts an associated `.yaml` in the `mcloud` folder.
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/README.md:14]: "an" -> "and"
"`.yaml`" -> "`.yaml` files"

File: examples/llm/inference/benchmarking/yamls/7b.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  vocab_size: 50368
+  init_std: 0.02
+  attn_pdrop: 0.0
+  resid_pdrop: 0.0
+  emb_pdrop: 0.0
+  attn_impl: triton
+
+fsdp_config:
+  sharding_strategy: FULL_SHARD
+  mixed_precision: DEFAULT
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/yamls/7b.yaml:32]: We're using mixed precision PURE now, yeah?

Comment (dakinggg) [reply examples/llm/inference/benchmarking/yamls/7b.yaml:32]: Actually why do these yamls need fsdp configs?

File: examples/llm/inference/benchmarking/yamls/7b.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  activation_checkpointing: false
+  activation_cpu_offload: false
+  verbose: false
+
+load_model_path: MosaicGPT.pt
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/yamls/7b.yaml:37]: Probably just leave this blank?

File: examples/llm/inference/benchmarking/mcloud/benchmark_launcher.yaml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+cluster: r0z0 # Update with your cluster here!
+image: mosaicml/composer:0.13.2
+integrations:
+- integration_type: git_repo
+  git_repo: mosaicml/examples
+  pip_install: '.[llm]'
+
+- integration_type: git_repo
+  git_repo: mosaicml/composer
+  # Right now composer 0.13.2 might not have the right HF changes right now
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/mcloud/benchmark_launcher.yaml:12]: What's going on here?

File: examples/llm/inference/benchmarking/mcloud/benchmark_launcher.yaml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  # So this is the current commit hash
+  git_commit: 0c704d31ead5963f57b3d5361a19666937a7e2d9
+  pip_install: '.[all]'
+
+command: |
+  cd examples/examples/llm/inference/benchmarking/
+
+  # For now since we are in the single GPU setting
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/mcloud/benchmark_launcher.yaml:20]: Can we just drop this comment and make sure we note somewhere (Jira) whatever the issue is? The comment right now is pretty confusing for a user. A new comment can just say this is single GPU for now.

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+import time
+
+import deepspeed
+import numpy as np
+import torch
+# You can use this to load the model weights
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:10]: Comment seems misplaced?

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+from composer.core import get_precision_context
+from omegaconf import OmegaConf as om
+from transformers import AutoTokenizer
+
+from examples.llm.src import COMPOSER_MODEL_REGISTRY
+
+
+def main(config):
+    inf_config = {
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:19]: "inf_config" -> "inference_config"

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        'tensor_parallel': {
+            'tp_size': 0
+        },
+    }
+
+    model = COMPOSER_MODEL_REGISTRY[config.model.name](config.model,
+                                                       config.tokenizer)
+    model.eval()
+
+    tokenizer = AutoTokenizer.from_pretrained(config.tokenizer_name)
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:33]: The tokenizer should be accessible via `model.tokenizer`

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py (hunk 3 above)
Comment (dakinggg) [reply examples/llm/inference/benchmarking/benchmark_deepspeed.py:39]: Can we just check if dtype is as expected here and raise exception if not? No need for prints i think

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+                times = []
+                eos_token = tokenizer.eos_token
+                # Make sure we are not generating a fake batch with a EOS token
+                while True:
+                    batch = torch.randint(
+                        0,
+                        config.model.vocab_size - 1,
+                        size=(batch_size, input_length
+                             )).to(f'cuda:{torch.cuda.current_device()}')
+                    if tokenizer.convert_tokens_to_ids(eos_token) not in batch:
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:56]: You can just use eos_token_id

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        print('dtype is: ', p.dtype)
+        break
+
+    stats = []
+    for batch_size in config.batch_sizes:
+        for input_length in config.input_lengths:
+            for output_length in config.output_lengths:
+                times = []
+                eos_token = tokenizer.eos_token
+                # Make sure we are not generating a fake batch with a EOS token
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:49]: I think it's actually ok to have an eos in the input, but anyway,can we just replace eos with something else rather than a (theoretically) infinite while loop?

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+                        size=(batch_size, input_length
+                             )).to(f'cuda:{torch.cuda.current_device()}')
+                    if tokenizer.convert_tokens_to_ids(eos_token) not in batch:
+                        break
+                batch = batch.to(torch.long)
+                for i in range(config.num_runs + 1):
+                    start_time = time.time()
+                    with torch.no_grad():
+                        if config.use_precision_context:
+                            with get_precision_context(config.precision):
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:63]: Can we use `context = get_precision_context if config.use_precision_context else contextlib.nullcontext()` instead?

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+                                    max_new_tokens=output_length,
+                                    use_cache=True)
+                        else:
+                            inference_model.generate(
+                                batch,
+                                max_new_tokens=output_length,
+                                use_cache=True)
+                    # We noticed there sometimes might be a small bit of startup time
+                    # so we only start to benchmark after the first iteration
+                    if i > 0:
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:75]: @vchiley is this enough warmup for triton benchmarking?

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+                        times.append(time.time() - start_time)
+
+                num_output_tokens = output_length * batch_size
+                mean_time = np.mean(times)
+                tokens_per_second = num_output_tokens / float(mean_time)
+
+                resu = (
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:82]: "resu" -> "result"

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+                             )).to(f'cuda:{torch.cuda.current_device()}')
+                    if tokenizer.convert_tokens_to_ids(eos_token) not in batch:
+                        break
+                batch = batch.to(torch.long)
+                for i in range(config.num_runs + 1):
+                    start_time = time.time()
+                    with torch.no_grad():
+                        if config.use_precision_context:
+                            with get_precision_context(config.precision):
+                                inference_model.generate(
```
Comment (dakinggg) [inline examples/llm/inference/benchmarking/benchmark_deepspeed.py:64]: Can you pass in the full attention mask here? Just to be safe

**Review by dakinggg — APPROVED** [review]
Handful of nitpicky comments, but generally lgtm!

**Comment by abhi-mosaic** [comment]
@bcui19 I think we should have a quick sync on what inference workflow we want to encourage here. I think we should refactor these scripts to load _HF models only_, not Composer checkpoints, and benchmarking should support either  either raw HF generate, or the DeepSpeed inference wrapper. I don't think we want to reference any training YAMLs or ComposerMosaicGPT models when we get to inference time.

Basically the export flow I want to encourage is: (Training YAML + Composer ckpt) -> (HF folder with model and tokenizer inside) -> optionally ONNX.

Checkout this JIRA for more details: https://mosaicml.atlassian.net/browse/RESEARCH-589

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py (hunk 15 above)
Comment (bcui19) [reply examples/llm/inference/benchmarking/benchmark_deepspeed.py:75]: I made it a flag. I think I remember the biggest gap was with the very first batch FWIW

File: examples/llm/inference/benchmarking/yamls/7b.yaml (hunk 6 above)
Comment (bcui19) [reply examples/llm/inference/benchmarking/yamls/7b.yaml:37]: I think for now I'm just going to remove...

File: examples/llm/inference/benchmarking/benchmark_deepspeed.py (hunk 15 above)
Comment (dskhudia) [reply examples/llm/inference/benchmarking/benchmark_deepspeed.py]: Yup. Having a flag is better. I made it a flag in my local version as well.

File: examples/llm/inference/benchmarking/benchmark.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+
+    if config.use_deepspeed:
+        model = deepspeed.init_inference(model, config=inference_config)
+
+        # Checking if deepspeed casts dtypes correctly
+        for _, p in model.named_parameters():
+            compare_precision(config.precision, p.dtype)
+            break
+    else:
+        model.to(torch.cuda.current_device())
```
Comment (abhi-mosaic) [inline examples/llm/inference/benchmarking/benchmark.py:52]: I think we want to cast the model to dtype here?

File: examples/llm/inference/benchmarking/benchmark.py [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+
+                # Make sure we are not generating a fake batch with a EOS token
+                eos_mask = batch == tokenizer.eos_token_id
+                batch = batch.masked_fill(eos_mask,
+                                          config.tokenizer.non_eos_token_id)
+
+                batch = batch.to(torch.long)
+
+                attention_mask = torch.logical_not(
+                    torch.eq(batch, tokenizer.eos_token_id))
```
Comment (abhi-mosaic) [inline examples/llm/inference/benchmarking/benchmark.py:81]: I'm not sure if this section is doing the right thing. We only want to avoid padding tokens, not EOS tokens. We can also remove the EOS token id in the generate_kwargs and that will force it to generate the `max_len` tokens every time.

File: examples/llm/inference/benchmarking/benchmark.py [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+                        with precision_context:
+                            model.generate(batch,
+                                           max_new_tokens=output_length,
+                                           use_cache=True,
+                                           attention_mask=attention_mask)
+
+                    # We noticed there sometimes might be a small bit of startup time
+                    # so we only start to benchmark after some number of batches
+                    if i >= config.num_warmup_batches:
+                        times.append(time.time() - start_time)
```
Comment (abhi-mosaic) [inline examples/llm/inference/benchmarking/benchmark.py:102]: you need to synchronize before every timestamp

File: examples/llm/inference/benchmarking/mcloud/benchmark_launcher.yaml [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+  # So this is the current commit hash as of 04/17/2023
+  # We will remove this install when we have a release
+  # with the appropriate changes.
+  git_commit: 20fb86b476d91749528a989265c3e211eb1d774d
+  pip_install: '.[all]'
+
+command: |
+  cd examples/examples/llm/inference/benchmarking/
+
+  python benchmark_deepspeed.py yamls/1b.yaml
```
Comment (abhi-mosaic) [inline examples/llm/inference/benchmarking/mcloud/benchmark_launcher.yaml:22]:
```suggestion
  python benchmark.py yamls/1b.yaml
```

File: examples/llm/inference/benchmarking/yamls/7b.yaml [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+
+batch_sizes: [1, 2, 4, 8, 16, 32, 64]
+input_lengths: [128]
+output_lengths: [8]
+num_runs: 5
+
+num_warmup_batches: 3
+
+use_precision_context: False
+use_deepspeed: True
```
Comment (abhi-mosaic) [inline examples/llm/inference/benchmarking/yamls/7b.yaml:39]: I think it would make sense to rename  the `precision` and `use_precision_context` flags.

Perhaps like:

```
param_dtype: [fp32, bf16, fp16]
autocast: [none, amp_bf16, amp_fp16]
```

Right now, the setups I would like to test are not easy to configure:
* bf16 weights, no autocast
* bf16 weights, autocast amp_bf16 (which will upcast certain ops to fp32)
* fp32 weights, autocast amp_bf16 (expected behavior is identical to above)

Leaving the naming and stuff to you, but whatever convention you decide I will update `hf_generate.py` to match.

File: examples/llm/inference/benchmarking/benchmark.py [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+
+                result = (
+                    f'{config.benchmark_name}_{batch_size}_{input_length}_{output_length}',
+                    f'{mean_time:.3f}', f'{tokens_per_second:.3f}')
+
+                run_name, latency, tokens_per_second = result
+
+                print(f'{run_name}, {latency}, {tokens_per_second}')
+
+                stats.append(result)
```
Comment (abhi-mosaic) [inline examples/llm/inference/benchmarking/benchmark.py:118]: It might be worth fleshing this out with a couple more values like `ms_per_seq_output_token` which is closer to the relevant user metric "how long do *I* have to wait for eac of *my* tokens when I'm lumped in with everyone else in a batch". Maybe like the rows of the 'inference-perf' google sheet.

And maybe adding strings to the `result` like `f'latency={mean_time:.3f}, ...'` because otherwise it'll look like a bunch of numbers and hard to remember what each one is

**Review by abhi-mosaic — COMMENTED** [review]
Left some comments, mostly around dtype / autocast options and metric reporting. Thank you @bcui19 !

File: examples/llm/inference/benchmarking/benchmark.py (hunk 23 above)
Comment (bcui19) [reply examples/llm/inference/benchmarking/benchmark.py:118]: Ah so right now I kept it display sth like:

Run name, latency (s), tokens per second
===========================================================================
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_1_128_8, 0.081, 99.302
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_2_128_8, 0.081, 196.963
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_4_128_8, 0.082, 392.072
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_8_128_8, 0.087, 731.439
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_16_128_8, 0.100, 1285.682
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_32_128_8, 0.128, 2007.501
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_64_128_8, 0.226, 2267.875

to make it easy to copy and paste into the google spreadsheet. Happy to change as well though. 

I will add `ms_per_seq_output_token` btw

Comment (bcui19) [reply examples/llm/inference/benchmarking/benchmark.py]: Actually nvm I see how to amend this. All good!

File: examples/llm/inference/benchmarking/yamls/1b.yaml [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+  max_seq_len: ${max_seq_len}
+  vocab_size: 50368
+  init_std: 0.02
+  attn_pdrop: 0.0
+  resid_pdrop: 0.0
+  emb_pdrop: 0.0
+  attn_impl: triton
+
+autocast_precision: amp_bf16
+model_dtype: amp_bf16
```
Comment (abhi-mosaic) [inline examples/llm/inference/benchmarking/yamls/1b.yaml:30]:
```suggestion
model_dtype: bf16
```

File: examples/llm/inference/benchmarking/yamls/7b.yaml [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+  max_seq_len: ${max_seq_len}
+  vocab_size: 50368
+  init_std: 0.02
+  attn_pdrop: 0.0
+  resid_pdrop: 0.0
+  emb_pdrop: 0.0
+  attn_impl: triton
+
+autocast_precision: amp_bf16
+model_dtype: amp_bf16
```
Comment (abhi-mosaic) [inline examples/llm/inference/benchmarking/yamls/7b.yaml:30]:
```suggestion
model_dtype: bf16
```

**Review by abhi-mosaic — APPROVED** [review]
Looks great! Last minor comments but approving to unblock.
